synapse/events/snapshot.py

```diff
@@ -254,6 +278,225 @@ async def get_prev_state_ids(
+        state_group: Optional[int],
+        state_delta_due_to_event: dict,
+        prev_group: Optional[int],
+        delta_ids: Optional[StateMap[str]],
+        partial_state: bool,
+        state_map: StateMap[str],
+        state_group_before_event: int,
+    ) -> None:
+        super().__init__(storage_controller)
+        self.state_group = state_group
+        self.state_delta_due_to_event = state_delta_due_to_event
+        self.prev_group = prev_group
+        self.delta_ids = delta_ids
+        self.partial_state = partial_state
+        self.state_map = state_map
+        self.state_group_before_event = state_group_before_event
+
+    async def get_prev_state_ids(
+        self, state_filter: Optional["StateFilter"] = None
+    ) -> StateMap[str]:
+        """
+        Gets the room state map, excluding this event.
+
+        Args:
+            state_filter: specifies the type of state event to fetch from DB, example: EventTypes.JoinRules
+
+        Returns:
+            Maps a (type, state_key) to the event ID of the state event matching
+            this tuple.
+        """
+        return self.state_map
+
+    async def persist(self, event: EventBase) -> EventContext:
+        """
+        If the event is a state event, calculates the current state group for the context,
+        stores it, and returns a EventContext. If the event is not state, returns
+        an EventContext.
+
+        Args:
+             event: event that the EventContext is associated with.
+
+        Returns: An EventContext suitable for sending to the database with the event
+        for persisting
+        """
+        assert self.partial_state is not None
+        if not event.is_state():
+            return EventContext.with_state(
+                storage=self._storage,
+                state_group=self.state_group_before_event,
+                state_group_before_event=self.state_group_before_event,
+                state_delta_due_to_event=self.state_delta_due_to_event,
+                partial_state=self.partial_state,
+                prev_group=self.prev_group,
+                delta_ids=self.delta_ids,
+            )
+        else:
+            state_group_after_event = await self._storage.state.store_state_group(
+                event.event_id,
+                event.room_id,
+                prev_group=self.prev_group,
+                delta_ids=self.delta_ids,
+                current_state_ids=None,
+            )
+
+            return EventContext.with_state(
+                storage=self._storage,
+                state_group=state_group_after_event,
+                state_group_before_event=self.state_group_before_event,
+                state_delta_due_to_event=self.state_delta_due_to_event,
+                partial_state=self.partial_state,
+                prev_group=self.prev_group,
+                delta_ids=self.delta_ids,
+            )
+
+
+class UnpersistedEventContext(UnpersistedEventContextBase):
```

Comment:
Yeah, these classes look identical except for the optional `state_map`. I'd merge them.